website/static/js/setPassword.js

```diff
@@ -0,0 +1,234 @@
+'use strict';
+
+var ko = require('knockout');
+require('knockout.validation');
+var $ = require('jquery');
+var zxcvbn = require('zxcvbn');
+
+var $osf = require('./osfHelpers');
+
+
+ko.validation.rules.complexity = {
+    validator: function (val, minimumComplexity) {
+        return zxcvbn(val).score >= minimumComplexity;
+    },
+    message: 'Please enter a more complex password.'
+};
+
+ko.validation.registerExtenders();
+
+
+// Accepts a few different types of arguments, depending on
+// the desired fields and type of password reset.
+// types:
+//    - signup: required fields for Name, email, email confirmation, password, and password strength
+//    - reset: fields for password, strength, and password confirmation (for when you don't know your old password)
+//    - change: change your password when you know your old one
+var ViewModel = function(passwordViewType, submitUrl, campaign, redirectUrl) {
+
+    var self = this;
+
+    self.typedPassword = ko.observable('');
+
+    self.passwordInfo = ko.computed(function() {
+        if (self.typedPassword()) {
+            return zxcvbn(self.typedPassword().slice(0, 100));
+        }
+    });
+
+    self.passwordFeedback = ko.computed(function () {
+        if (self.typedPassword()) {
+            return self.passwordInfo().feedback.warning;
+        }
+    });
+
+    self.passwordComplexity = ko.pureComputed(function() {
+        if (self.typedPassword()) {
+            return self.passwordInfo().score;
+        }
+    });
+
+    self.passwordComplexityBar = ko.computed(function() {
+        return $osf.valueProgressBar(self.passwordComplexity());
+    });
+
+    self.password = ko.observable('').extend({
+        required: true,
+        minLength: 6,
+        maxLength: 256,
+        complexity: 2,
+    });
+
+    self.campaign = ko.observable(campaign);
+
+    // Preserve object of validated fields for use in `submit`
+    var validatedFields = {
+        password: self.password
+    };
+
+    if (passwordViewType === 'reset' || passwordViewType === 'change') {
```

Comment:
Just a thought: it would be cool to refactor out this page-specific (passwordViewType) code. I can think of two ways.

1) Add a way to extend an observable with password-checking, a la knockout-validation. Something like

```js
var password = ko.observable('').extend({passwordChecking: true});
//...
// You would access the below methods in the HTML
password.passwordComplexity() // => 3
password.passwordSuggestions() // ['Too easy', ...]
```

```html
<p data-bind="foreach: password.passwordSuggestions()"> ...
```

I think this is the ideal API, as it allows very easy reuse of password validation logic. I haven't given much thought to implementation, though.

Alternatively...

2) Add a custom binding handler


```html
<input type="password" data-bind="passwordValidation: {observable: password, messages: pwMessages, complexity: pwComplexity}" />
```

```js
function ViewModel(..) {
    self.password = ko.obs();
    self.pwMessages = ko.obsArray();
    self. pwComplexity = ko.obs();
}
```

This API is less ergonomic than 1), but *might* be more straightforward to implement.


These are just thoughts, though. Refactoring isn't a blocker to get this merged, but it's worth investigating.